smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/commit/KafkaFileCheckpointCommit.java

```diff
@@ -0,0 +1,103 @@
+package io.smallrye.reactive.messaging.kafka.commit;
+
+import java.io.File;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.TopicPartition;
+
+import io.smallrye.common.annotation.Experimental;
+import io.smallrye.common.annotation.Identifier;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.reactive.messaging.kafka.KafkaConnectorIncomingConfiguration;
+import io.smallrye.reactive.messaging.kafka.KafkaConsumer;
+import io.vertx.core.json.Json;
+import io.vertx.mutiny.core.Vertx;
+import io.vertx.mutiny.core.buffer.Buffer;
+
+/**
+ * Checkpointing commit handler which uses local files as state store. It creates a file per topic-partition.
+ */
+@Experimental("Experimental API")
+public class KafkaFileCheckpointCommit extends KafkaCheckpointCommit {
+
+    private final Vertx mutinyVertx;
+    private final File stateDir;
+
+    public KafkaFileCheckpointCommit(KafkaConnectorIncomingConfiguration config,
+            Vertx vertx,
+            KafkaConsumer<?, ?> consumer,
+            BiConsumer<Throwable, Boolean> reportFailure,
+            int defaultTimeout,
+            File stateDir) {
+        super(vertx, config, consumer, reportFailure, defaultTimeout);
+        this.mutinyVertx = vertx;
+        this.stateDir = stateDir;
+    }
+
+    @ApplicationScoped
+    @Identifier("checkpoint-file")
+    public static class Factory implements KafkaCommitHandler.Factory {
+
+        @Override
+        public KafkaCommitHandler create(KafkaConnectorIncomingConfiguration config, Vertx vertx,
+                KafkaConsumer<?, ?> consumer, BiConsumer<Throwable, Boolean> reportFailure) {
+            int defaultTimeout = config.config()
+                    .getOptionalValue(ConsumerConfig.DEFAULT_API_TIMEOUT_MS_CONFIG, Integer.class)
+                    .orElse(60000);
+            String stateDir = config.config().getValue("checkpoint-state-file", String.class);
+
+            return new KafkaFileCheckpointCommit(config, vertx, consumer, reportFailure, defaultTimeout,
+                    new File(stateDir));
+        }
+    }
+
+    private String getStatePath(TopicPartition partition) {
+        return stateDir.toPath().resolve(partition.topic() + "-" + partition.partition()).toString();
+    }
+
+    @Override
+    protected Uni<ProcessingState<?>> fetchProcessingState(TopicPartition partition) {
+        String statePath = getStatePath(partition);
+        return mutinyVertx.fileSystem().exists(statePath).chain(exists -> {
+            if (exists)
+                return mutinyVertx.fileSystem().readFile(statePath).map(this::deserializeState)
```

Comment:
So... File systems. They are not truly async (until we have io_uring). It dispatches the work on an internal worker thread. Not a big deal here, but better to keep that in mind.

Reply:
Yeah, it is useful to test the feasibility of the checkpointing API but, for sure this implementation is not production-ready.